### PR TITLE
32 staking contracts should earn rewards

### PR DESCRIPTION
**Pull request by aliberkhsl**

### CODE REVIEW

File: contracts/staking/BaseShareField.sol [hunk 1]
```diff
@@ -16,28 +31,48 @@ contract BaseShareField {
...
         if (totalProductivity == 0) {
-            totalShare = totalShare.add(_currentReward());
+            lastRewardBlock = block.number;
             return;
         }
+        FundManager fm = FundManager(nameService.getAddress("FUND_MANAGER"));
+        (uint rewardsPerBlock, uint blockStart, uint blockEnd, bool isBlackListed) = fm.getStakingReward(address(this));
+        if(block.number >= blockStart && blockEnd>= block.number && isBlackListed == false){
+            uint256 multiplier = block.number.sub(lastRewardBlock);
+            uint256 reward = multiplier.mul(rewardsPerBlock * 10 ** 16); // turn it to 18 decimals
```
Comment (gooddollar-techadmin) [inline contracts/staking/BaseShareField.sol:69]: \* 1e16

Comment (gooddollar-techadmin) [reply contracts/staking/BaseShareField.sol:69]: no need for .mul .add etc in solidity 8

File: contracts/staking/BaseShareField.sol [hunk 2]
```diff
@@ -16,28 +31,48 @@ contract BaseShareField {
...
+
+            accAmountPerShare = accAmountPerShare.add(reward.mul(1e12).div(totalProductivity));
```
Comment (gooddollar-techadmin) [inline contracts/staking/BaseShareField.sol:71]: whilee mul 1e12? always add comments explaining math

File: contracts/staking/BaseShareField.sol [hunk 3]
```diff
@@ -16,28 +31,48 @@ contract BaseShareField {
...
+            totalShare = totalShare.add(reward);
```
Comment (gooddollar-techadmin) [inline contracts/staking/BaseShareField.sol:72]: what is totalShare? is it like the total rewards this contract has earned  to be divided by  stakers? maybe totalRewards is a better name?

File: contracts/staking/BaseShareField.sol [hunk 4]
```diff
@@ -48,10 +83,22 @@ contract BaseShareField {
     function _audit(address user) internal virtual {
         UserInfo storage userInfo = users[user];
         if (userInfo.amount > 0) {
+            uint256 blocksPaid = userInfo.lastRewardTime.sub(userInfo.multiplierResetTime);
```
Comment (gooddollar-techadmin) [inline contracts/staking/BaseShareField.sol:86]: add comment // lastRewardTime is always >= multiplierResetTime

File: contracts/staking/BaseShareField.sol [hunk 5]
```diff
@@ -48,10 +83,22 @@ contract BaseShareField {
...
+            uint256 blocksPassedFirstMonth = Math.min(172800,block.number.sub(userInfo.multiplierResetTime)); //172800 is equivalent of month in blocks
```
Comment (gooddollar-techadmin) [inline contracts/staking/BaseShareField.sol:87]: 172800 should be a defined as a constant

File: contracts/staking/BaseShareField.sol [hunk 6]
```diff
@@ -48,10 +83,22 @@ contract BaseShareField {
...
+            uint256 blocksPaid = userInfo.lastRewardTime.sub(userInfo.multiplierResetTime);
+            uint256 blocksPassedFirstMonth = Math.min(172800,block.number.sub(userInfo.multiplierResetTime)); //172800 is equivalent of month in blocks 
+            uint256 blocksToPay = block.number.sub(userInfo.lastRewardTime);
+            uint256 firstMonthBlocksToPay = blocksPaid >= (172800) ? 0 : blocksPassedFirstMonth.sub(blocksPaid);
+            uint256 fullBlocksToPay = blocksToPay.sub(firstMonthBlocksToPay);
+            
             uint pending = userInfo.amount.mul(accAmountPerShare).div(1e12).sub(userInfo.rewardDebt);
+           
+            if (blocksToPay != 0){
+                uint rewardPerBlock = pending.mul(uint256(10 ** 12)).div(blocksToPay).div(uint256(10 ** 12)); // increase resolution
```
Comment (gooddollar-techadmin) [inline contracts/staking/BaseShareField.sol:95]: use e notations

File: contracts/staking/BaseShareField.sol [hunk 7]
```diff
@@ -48,10 +83,22 @@ contract BaseShareField {
...
+                pending  = ((firstMonthBlocksToPay.mul(50*10**18).div(100)) + fullBlocksToPay).mul(rewardPerBlock).div(1e18); // divide 1e16 so reduce it to 18decimals
```
Comment (gooddollar-techadmin) [inline contracts/staking/BaseShareField.sol:96]: comment says 1e16 but you divide 1e18
better comments required on math

Comment (gooddollar-techadmin) [reply contracts/staking/BaseShareField.sol:96]: explain the calculation you are doing,
even maybe add in thee function comments the formulas from the word file

File: contracts/staking/GoodFundManager.sol [hunk 8]
```diff
@@ -60,7 +63,15 @@ contract GoodFundManager is DAOContract {
...
-
+    //Structure that hold reward information and if its blacklicksted or not for particular staking Contract
+    struct Reward{
+        uint32 blockReward; //in G$
+        uint32 blockStart;
+        uint32 blockEnd;
+        bool isBlackListed;
+    }
+    // Rewards per block for particular Staking contract
+    mapping(address => Reward) rewardsForStakingContract;
```
Comment (gooddollar-techadmin) [inline contracts/staking/GoodFundManager.sol:74]: in general make variables public unless they dont hold any interesting or trivial information

File: contracts/staking/GoodFundManager.sol [hunk 9]
```diff
@@ -138,7 +149,32 @@ contract GoodFundManager is DAOContract {
...
+        uint32 _blockEnd
+    ) public onlyAvatar{
+        Reward memory reward = Reward(_rewardsPerBlock, _blockStart, _blockEnd, false);
+        rewardsForStakingContract[_stakingAddress] = reward;
+    }
+    /**
+     * @dev Get the Reward for particular staking contract
+     * @param _stakingAddress address of the staking contract
+     */
+     function getStakingReward(
```
Comment (gooddollar-techadmin) [inline contracts/staking/GoodFundManager.sol:172]: no need for this  meetthod
just make it public per comment above

File: test/staking/StakingRewards.test.ts [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    await ictrl.genericCall(goodFundManager.address, encodedDataTwo, avatar, 0);
+    await setDAOAddress("MARKET_MAKER", marketMaker.address);
+    await setDAOAddress("FUND_MANAGER", goodFundManager.address);
+  });
+
+  it("should get g$ minting permissions", async () => {
+    expect(await goodReserve.dao()).to.be.equal(controller);
+    expect(await goodReserve.avatar()).to.be.equal(avatar);
+    await goodReserve.start();
+  });
```
Comment (gooddollar-techadmin) [inline test/staking/StakingRewards.test.ts:184]: dont use tests to initialize stuff

File: test/staking/StakingRewards.test.ts [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+  it("should be set fundmanager address in NameService", async () => {
+    let fundManagerAddress = await nameService.getAddress("FUND_MANAGER");
+    expect(fundManagerAddress).to.be.equal(goodFundManager.address);
+  });
```
Comment (gooddollar-techadmin) [inline test/staking/StakingRewards.test.ts:189]: not relevant

File: test/staking/StakingRewards.test.ts [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+      goodCompoundStaking.address
+    );
+    expect(rewardPerBlock[0].toString()).to.be.equal("1000");
+    expect(rewardPerBlock[1].toString()).to.be.equal("0")
+    expect(rewardPerBlock[2].toString()).to.be.equal("100")
+    expect(rewardPerBlock[3]).to.be.equal(false)
+    
+  });
+
+  it("should be able to earn rewards after some block passed", async () => {
```
Comment (gooddollar-techadmin) [inline test/staking/StakingRewards.test.ts:202]: we need much more tests here to check all aspects of our rewards calculation
you can also copy tests from baseshare etc

**Review by gooddollar-techadmin — CHANGES_REQUESTED** [review]

File: contracts/staking/BaseShareField.sol (hunk 3 above)
Comment (aliberkhsl) [reply contracts/staking/BaseShareField.sol:72]: No needed in our implementation so I deleted

File: test/staking/StakingRewards.test.ts (hunk 12 above)
Comment (aliberkhsl) [reply test/staking/StakingRewards.test.ts:202]: I'll add more when cancel out old interest mechanism after #39

**Comment by sirpy** [comment]
included in #41

**Pull request closed without merging**